### PR TITLE
Log in and Registration Test

### PR DESCRIPTION
**Pull request by dusan011**
I wrote log in and registration test. For registration I had a problem with selecting a country from drop down

### CODE REVIEW

File: cypress.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "$schema": "https://on.cypress.io/cypress.schema.json",
-  "baseUrl": "https://example.cypress.io",
+  "baseUrl": "https://develop--dib-travel.netlify.app/",
```
Comment (sjova) [inline cypress.json:3]: A similar change in the future should be discussed with a whole team.

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+describe('Login Test', () => {
+  let details: any;
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts:2]: We should add TypeScript interface (data model) for **details** type.

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  before(() => {
+    cy.fixture('loginData').then((data) => {
+      details = data;
+    });
+  });
+  beforeEach(() => {
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts:8]: For better readability, we should separate each block with a new line.

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    cy.get('.form-wrapper').within(() => {
+      cy.get('input[name="email"]').type(details.validEmail);
+      cy.get('input[name="password"]').type(details.validPassword);
+      cy.get('button').contains('Sign in').click();
+    });
+    cy.dataCy('navbar-my-travels-link').should('contain', 'My Travels');
+    cy.get('dib-hamburger-icon').click();
+    cy.get('dib-navbar-panel').should('contain', 'Log Out');
+  });
+  it.only('Negative Login Test', () => {
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts:28]: `*.only` - you should double-check the whole code before final PR

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      cy.get('[placeholder="Code"]').click();
+      //cy.get('.cdk-overlay-pane').contains(details.codeCountry).click() //did not know how to select specific country from a dropdown
+      cy.get('input[type="number"]').type(details.phoneNumber);
+      cy.get('[name="password"]').eq(0).type(details.password);
+      cy.get('[name="password"]').eq(2).type(details.password);
+      cy.get('[name="companyName"]').type(details.companyName);
+      cy.get('[name="companyRegistrationNumber"]').type(details.companyRegistrationNumber);
+
+      cy.get('ui-button button').click();
+      //cy.wait(3000);
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts:27]: We should skip `cy.wait` whenever is possible. Usually, we can attach action when the specific API call is done.

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  beforeEach(() => {
+    cy.visit('/sign-up');
+  });
+  it('Positive Registration Test', () => {
+    cy.get('.auth-container-content').within(() => {
+      cy.get('input[name="firstName"]').type(details.firstName);
+      cy.get('input[name="lastName"]').type(details.lastName);
+      cy.get('input[name="email"]').type(details.email);
+      cy.get('[placeholder="Code"]').click();
+      //cy.get('.cdk-overlay-pane').contains(details.codeCountry).click() //did not know how to select specific country from a dropdown
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts:19]: Please add `// TODO: some comment` block for any unfinished code in the future.

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    cy.visit('https://develop--dib-travel.netlify.app/login');
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts:29]: Why did you add the whole URL here?

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import { type } from 'cypress/types/jquery';
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts:1]: Why you added this block: `import { type } from 'cypress/types/jquery';` ?

Comment (dusan011) [reply cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts:1]: It was automatically added when I wanted to use type command but I have not seen it. My mistake

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts (hunk 7 above)
Comment (dusan011) [reply cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts:29]: My mistake. I can only write '/login' and it works. Is that better approach?

Comment (sjova) [reply cypress/integration/dib-playground/dusan.petrovic_tests/dib_login.spec.ts:29]: Yes, that is a correct approach when you define `baseUrl` in the config file.

File: cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts (hunk 8 above)
Comment (sjova) [reply cypress/integration/dib-playground/dusan.petrovic_tests/dib_registration.spec.ts:1]: Funny thing, included line is deprecated :)

File: cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import { Login } from '../../../../models';
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts:1]: 😍

File: cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+describe('Login Test', () => {
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts:3]: You don't need `* Test` suffix here, since we're in the test file (`*.spec.ts`) :)

You can only add a more descriptive description. Ex. **Login Page** or something.

As a reference you can use: https://github.com/cypress-io/cypress-realworld-app/tree/develop/cypress

File: cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  let details: Login;
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts:4]: It's nice to add more context in a variable name. Ex. `let loginDetails: Login;` or `let loginTestCases: Login;` or something similar

File: cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    cy.fixture('login-data').then((data) => {
+      details = data;
+    });
+  });
+
+  beforeEach(() => {
+    cy.visit('/');
+  });
+
+  it('Empty Login Test', () => {
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts:16]: each `it` should better describe test.

Examples:

`should ... something`
`display something`
`navigates to ...`
`submits ...`
`shows something`
`creates something`

As a reference you can use: https://github.com/cypress-io/cypress-realworld-app/tree/develop/cypress

File: cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    cy.get('.form-wrapper').within(() => {
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/login/dib-login.spec.ts:17]: It's totally ok to use `.within`, but if whenever is possible stick with simple `.get` (this will give us faster test execution time).

File: cypress/integration/dib-playground/dusan.petrovic_tests/registration/dib-registration.spec.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    cy.fixture('registration-data').then((data) => {
+      details = data;
+    });
+  });
+
+  beforeEach(() => {
+    cy.visit('/sign-up');
+  });
+
+  it('Positive Registration Test', () => {
```
Comment (sjova) [inline cypress/integration/dib-playground/dusan.petrovic_tests/registration/dib-registration.spec.ts:16]: This test is failed. Please double-check.

**Review by sjova — COMMENTED** [review]
The latest changes look good. I added few more minor improvements (these will produce long-term better code).

Sidenote: You don't need to use `dib-*` prefix for file naming since we will group features properly on the parent level.

Also, you didn't pull the latest changes from the `main` branch.

And remains use better selectors, as we discussed/

---

As a reference you can use:
https://github.com/cypress-io/cypress-realworld-app/tree/develop/cypress
https://github.com/cypress-io/cypress-example-kitchensink/tree/master/cypress
https://github.com/cypress-io/cypress-example-recipes

File: cypress/integration/dib-playground/dusan.petrovic_tests/registration/dib-registration.spec.ts (hunk 14 above)
Comment (dusan011) [reply cypress/integration/dib-playground/dusan.petrovic_tests/registration/dib-registration.spec.ts]: yes, it failed because I used the same email address for registration. Probably, better approach should be to create a function that generates random email address (dusan123@gmail.com, dusan321@gmail.com) - something like that

Comment (sjova) [reply cypress/integration/dib-playground/dusan.petrovic_tests/registration/dib-registration.spec.ts]: For now, it's totally ok to use generated different email.

As we agreed, later we'll rethink the testing process in different environments and removing test data.

Advanced Task (Optional): Collect used test email addresses in a single JSON file, so BE can easily remove those test data later (Hink: If you start to complicate things, you're on wrong path :))
